File: source/python/EuclideanAlgorithm.py
#!/usr/bin/python
# -*- coding: utf-8 -*-


"""
There are two implementations of the Euclidean algorithm, one iterative and one 
recursive, which is used to calculate the Greatest Common Divisor (GCD) of two 
numbers.
"""


__author__ = "Tommaso Tamagnini"
__copyright__ = "Copyright 2019, Tommaso Tamagnini"
__license__ = "MIT License"
__version__ = "1.0"


def iterative_gcd(n, m):
    """
    Calculates the Greatest Common Divisor of the two specified numbers, in an 
    iterative way.

    :param n: The first number to calculate the GCD
    :type n: int
    :param m: The second number to calculate the GCD
    :type m: int
    :returns: The GCD
    :rtype: int
    """

    while m > 0: (n, m) = (m, n % m)
    return n


def recursive_gcd(n, m):
    """
    Calculates the Greatest Common Divisor of the two specified numbers, in a 
    recursive way.

    :param n: The first number to calculate the GCD
    :type n: int
    :param m: The second number to calculate the GCD
    :type m: int
    :returns: The GCD
    :rtype: int
    """

    if m == 0: return n
    else: return recursive_gcd(m, n % m)

File: source/c#/EuclideanAlgorithm.cs
namespace EuclideanAlgorithm
{
	/// <summary>
	/// The class implements two version of the Euclidean algorithm, one
	/// iterative and one recursive, which is used to calculate the Greatest
	/// Common Divisor (GCD) of two numbers.
	/// </summary>
	public static class EuclideanAlgorithm
	{
		/// <summary>
		/// Calculates the Greatest Common Divisor of the two specified numbers,
		/// in an iterative way.
		/// </summary>
		/// <param name="n">The first number to calculate the GCD</param>
		/// <param name="m">The second number to calculate the GCD</param>
		/// <returns>The GCD</returns>
		public static int IterativeGCD(int n, int m)
		{
			while (m > 0)
			{
				int temp = n;
				n = m;
				m = temp % m;
			}
			return n;
		}

		/// <summary>
		/// Calculates the Greatest Common Divisor of the two specified numbers,
		/// in a recursive way.
		/// </summary>
		/// <param name="n">The first number to calculate the GCD</param>
		/// <param name="m">The second number to calculate the GCD</param>
		/// <returns>The GCD</returns>
		public static int RecursiveGCD(int n, int m)
		{
			if (m == 0) return n;
			return RecursiveGCD(m, n % m);
		}
	}
}
